File: crates/defi-protocols/src/derivatives/hlp.rs
use alloy::primitives::{Address, U256};
use alloy::sol;
use alloy::sol_types::SolCall;
use async_trait::async_trait;

use defi_core::error::{DefiError, Result};
use defi_core::traits::Derivatives;
use defi_core::types::*;

sol! {
    #[sol(rpc)]
    interface IHlpVault {
        function deposit(uint256 amount) external returns (uint256);
        function withdraw(uint256 shares) external returns (uint256);
    }
}

pub struct HlpVault {
    name: String,
    vault: Address,
}

impl HlpVault {
    pub fn from_contracts(
        name: String,
        contracts: &std::collections::HashMap<String, Address>,
    ) -> Result<Self> {
        let vault = contracts
            .get("vault")
            .copied()
            .ok_or_else(|| DefiError::ContractError("Missing 'vault' contract".to_string()))?;
        Ok(Self { name, vault })
    }
}

#[async_trait]
impl Derivatives for HlpVault {
    fn name(&self) -> &str {
        &self.name
    }

    async fn build_open_position(&self, params: DerivativesPositionParams) -> Result<DeFiTx> {
        // HLP is a vault-style product, deposit = open position
        let call = IHlpVault::depositCall {
            amount: params.collateral,
        };
        Ok(DeFiTx {
            description: format!(
                "[{}] Deposit {} into HLP vault",
                self.name, params.collateral
            ),
            to: self.vault,
            data: call.abi_encode().into(),
            value: U256::ZERO,
            gas_estimate: Some(200_000),
        })
    }

    async fn build_close_position(&self, params: DerivativesPositionParams) -> Result<DeFiTx> {
        let call = IHlpVault::withdrawCall {
            shares: params.size,
        };
        Ok(DeFiTx {
            description: format!("[{}] Withdraw {} from HLP vault", self.name, params.size),
            to: self.vault,
            data: call.abi_encode().into(),
            value: U256::ZERO,
            gas_estimate: Some(200_000),
        })
    }
}

File: ts/packages/defi-protocols/src/derivatives/hlp.ts
import { parseAbi, encodeFunctionData } from "viem";
import type { IDerivatives } from "@hypurrquant/defi-core";
import {
  DefiError,
  type ProtocolEntry,
  type DerivativesPositionParams,
  type DeFiTx,
} from "@hypurrquant/defi-core";
import type { Address } from "viem";

const HLP_ABI = parseAbi([
  "function deposit(uint256 amount) external returns (uint256)",
  "function withdraw(uint256 shares) external returns (uint256)",
]);

export class HlpVaultAdapter implements IDerivatives {
  private readonly protocolName: string;
  private readonly vault: Address;

  constructor(entry: ProtocolEntry, _rpcUrl?: string) {
    this.protocolName = entry.name;
    const vault = entry.contracts?.["vault"];
    if (!vault) throw DefiError.contractError("Missing 'vault' contract");
    this.vault = vault;
  }

  name(): string {
    return this.protocolName;
  }

  async buildOpenPosition(params: DerivativesPositionParams): Promise<DeFiTx> {
    const data = encodeFunctionData({
      abi: HLP_ABI,
      functionName: "deposit",
      args: [params.collateral],
    });
    return {
      description: `[${this.protocolName}] Deposit ${params.collateral} into HLP vault`,
      to: this.vault,
      data,
      value: 0n,
      gas_estimate: 200_000,
    };
  }

  async buildClosePosition(params: DerivativesPositionParams): Promise<DeFiTx> {
    const data = encodeFunctionData({
      abi: HLP_ABI,
      functionName: "withdraw",
      args: [params.size],
    });
    return {
      description: `[${this.protocolName}] Withdraw ${params.size} from HLP vault`,
      to: this.vault,
      data,
      value: 0n,
      gas_estimate: 200_000,
    };
  }
}
